feat: add shared lockbox spec (#7)
defi-wonderland/specs: specs/interop/shared-lockbox.md

# Shared Lockbox

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [Design](#design)
  - [Interface and properties](#interface-and-properties)
  - [Events](#events)
- [Reference implementation](#reference-implementation)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

> With interoperable ETH, withdrawals will fail if the referenced `OptimismPortal` lacks sufficient ETH.

Reviewer: Assuming that lot of people will ready the spec without the design doc, this would be a bit too little context

The `SharedLockbox` improves the Superchain's interoperable ETH withdrawal user experience and avoids this issue.
To do so, it unifies ETH L1 liquidity in a single contract (`SharedLockbox`), enabling seamless withdrawals of ETH
from any OP chain in the Superchain, regardless of where the ETH was initially deposited.

## Design

The `SharedLockbox` contract is designed to manage the unified ETH liquidity for the Superchain.
It implements two main functions: `lockETH` for depositing ETH into the lockbox,
and `unlockETH` for withdrawing ETH from the lockbox.
These functions are called by the `OptimismPortal` contracts to manage the shared ETH liquidity

> when making deposits or finalizing withdrawals.

Reviewer: Should we add here something like "These `OptimismPortals` will be whitelisted by the SuperchanConfig using the `authorizePortal` function"

- The function MUST emit the `ETHLocked` event with the `portal` that called it and the `amount`.

```solidity
function lockETH() external payable;
```

**`unlockETH`**

Withdraws a specified amount of ETH from the lockbox's liquidity pool.

- Only authorized `OptimismPortal` addresses SHOULD be allowed to interact.
- The function MUST emit the `ETHUnlocked` event with the `portal` that called it and the `amount`.

```solidity
function unlockETH(uint256 _value) external;
```

**`authorizePortal`**

Grants authorization to a specific `OptimismPortal` contract.

- Only `SuperchainConfig` address SHOULD be allowed to interact.
- The function MUST add the specified address to the mapping of authorized portals.
- The function MUST emit the [`AuthorizedPortal`](#events) event when a portal is successfully added.

```solidity
function authorizePortal(address _portal) external;
```

### Events

**`ETHLocked`**

MUST be triggered when `lockETH` is called

```solidity
event ETHLocked(address indexed portal, uint256 amount);
```

**`ETHUnlocked`**

MUST be triggered when `unlockETH` is called

```solidity
event ETHUnlocked(address indexed portal, uint256 amount);
```

> **`AuthorizedPortal`**

Reviewer: To be consistent with the other events in this contract, i feel like it should be called `PortalAuthorized` since on the ETHLocked and ETHUnlocked we have `SUBJECT` first and `ADJECTIVE` later to describe the state of the subject. Here it would be the same "Portal" + "Authorized"

# Shared Lockbox

<!-- START doctoc generated TOC please keep comment here to allow auto update -->
<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [Design](#design)
  - [Contracts inroductions](#contracts-inroductions)
    - [`SharedLockbox`](#sharedlockbox)
    - [`LiquidityMigrator`](#liquiditymigrator)
  - [Contracts upgrades](#contracts-upgrades)
    - [`SuperchainConfig`](#superchainconfig)
    - [`OptimismPortal`](#optimismportal)
- [Reference implementation](#reference-implementation)
  - [`SharedLockbox`](#sharedlockbox-1)
  - [`LiquidityMigrator`](#liquiditymigrator-1)
  - [`SuperchainConfig`](#superchainconfig-1)
  - [`OptimismPortal`](#optimismportal-1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

> With interoperable ETH, withdrawals may fail if the referenced `OptimismPortal` lacks sufficient ETH

Reviewer 2: ```suggestion
With interoperable ETH, withdrawals will fail if the referenced `OptimismPortal` lacks sufficient ETH
```
nitpick to make the sentence more assertive

> — making it mandatory to withdraw from the L2 where the ETH was originally deposited.

Reviewer 2: ```suggestion
— making it mandatory to withdraw from the L2 where the ETH was originally deposited.
```

does this make sense? i don't fully get it, why would it be mandatory?

Reviewer 3: I think it refers to an alternative approach to the lockbox in which the ETH per L2 is tracked on the portal, is that right @agusduha?

But maybe is a good idea to remove this sentence, I think the paragraph is clear enough without it

The `SharedLockbox` improves the Superchain's interoperable ETH withdrawal user experience and avoid this issue.

> To do so, it unifies ETH L1 liquidity in a single contract, enabling seamless withdrawals of ETH from any OP chain

Reviewer 2: ```suggestion
To do so, it unifies ETH L1 liquidity in a single contract (`SharedLockbox`), enabling seamless withdrawals of ETH from any OP chain
```

if you want full clarity

in the Superchain, regardless of where the ETH was initially deposited.

## Design

> ### Contracts inroductions

Reviewer 2: ```suggestion
### Contracts introductions
```

#### `SharedLockbox`

The `SharedLockbox` contract is designed to manage the unified ETH liquidity for the Superchain.
It provides two main functions: `lockETH` for depositing ETH into the lockbox,
and `unlockETH` for withdrawing ETH from the lockbox.
These functions are called by the `OptimismPortal` contracts to manage the shared ETH liquidity
when making deposits or finalizing withdrawals.

**Interface and properties:**

**`lockETH`**

Deposits and locks ETH into the lockbox's liquidity pool.

- The function MUST accept ETH.

> - Only authorized `OptimismPortal` addresses MUST be allowed to interact.

Reviewer 3: wdyt? I find MUST more related to a specific action (like MUST emit ...) whereas in this case we're talking about permissions
```suggestion
- Only authorized `OptimismPortal` addresses SHOULD be allowed to interact.
```

> - The function MUST emit the `ETHLocked` event with the correct `portal` and `amount`.

Reviewer 2: Perhaps could be good to define what the "correct" portal is despite it being somewhat obvious for maximum clarity

```solidity
function lockETH() external payable;
```

**`unlockETH`**

Withdraws a specified amount of ETH from the lockbox's liquidity pool.

> - Only authorized `OptimismPortal` addresses MUST be allowed to interact.

Reviewer 3: Same here, I'm not sure if MUST is the correct keyword
```suggestion
- Only authorized `OptimismPortal` addresses SHOULD be allowed to interact.
```

- The function MUST emit the `ETHUnlocked` event with the correct `portal` and `amount`.

```solidity
function unlockETH(uint256 _value) external;
```

**`authorizePortal`**

Grants authorization to a specific `OptimismPortal` contract.

> - Only `SuperchainConfig` address MUST be allowed to interact.

Reviewer 3: Same doubt here
```suggestion
- Only `SuperchainConfig` address SHOULD be allowed to interact.
```

- The function MUST add the specified address to the mapping of authorized portals.
- The function MUST emit the `AuthorizedPortal` event when a portal is successfully added.

```solidity
function authorizePortal(address _portal) external;
```

**`ETHLocked`**

> MUST trigger when `lockETH` is called

Reviewer 3: ```suggestion
MUST be triggered when `lockETH` is called
```

Reviewer 3: If you agree, the same can be applied over the other events wording

```solidity

> event ETHLocked(address indexed portal, uint256 amount);

Reviewer 2: i know indexing by portal makes a ton of sense and is a good enough filter, but just for the sake of discussion, would there be any benefit to indexing the amount as well?

Reviewer 3: Maybe if in a case where you have an ETH withdrawal with always the same amount?
IMO is not worth it tho

Author: The amount comes from user different inputs so i dont see it being useful

```

**`ETHUnlocked`**

MUST trigger when `unlockETH` is called

```solidity
event ETHUnlocked(address indexed portal, uint256 amount);
```

**`AuthorizedPortal`**

MUST trigger when `authorizePortal` is called

```solidity
event AuthorizedPortal(address indexed portal);
```

> #### `LiquidityMigrator`

Reviewer 3: I find beneficial to explain why using this instead of just creating a one time function or executing this logic on `initialize()` inside the Portal?

> This contract is meant to be used as intermediate step for the liquidity migration.

Reviewer 3: ```suggestion
This contract is meant to be used as an intermediate step for the liquidity migration.
```

> It main purpose is to transfer the whole ETH balance from the `OptimismPortal` to the `SharedLockbox`.

Reviewer 3: ```suggestion
Its main purpose is to transfer the whole ETH balance from the `OptimismPortal` to the `SharedLockbox`.
```

Reviewer 3: Also, how do you feel about removing 'the' here?
```
from `OptimismPortal` to `SharedLockbox`.
```

Reviewer 3: Starting with 'Its unique purpose ... '?

Author: Sorry i missunderstood the comment haha, im totally fine with the change u suggested

**Interface and properties:**

**`migrateETH`**

> Transfers the entire ETH balance from the OptimismPortal to the SharedLockbox.

Reviewer 2: ```suggestion
Transfers the entire ETH balance from the `OptimismPortal` to the `SharedLockbox`.
```

- It MUST transfer the whole ETH balance to the `SharedLockbox` when called.

```solidity
function migrateETH() external;
```

### Contracts upgrades

#### `SuperchainConfig`

> This contract will be updated to manage and keep track the dependency graph.

Reviewer 2: ```suggestion
This contract will be updated to manage and keep track of the dependency graph.
```

It will be queried as the source of truth to get which chains are part of the Superchain.
It assumes the role of a `dependencyManager` per `SystemConfig` contract involved.
It will also allow to add a chain to the op-governed cluster and update each chain’s dependency set.

**Interface and properties**

The `SuperchainConfig` contract will add the following storage layout and function:

**`SHARED_LOCKBOX`**

- An immutable address pointing to the `SharedLockbox` contract.

> - This address MUST immutable because there's only one `SharedLockbox` for each cluster.

Reviewer 3: ```suggestion
- This address MUST be immutable because there's only one `SharedLockbox` for each cluster.
```

**`systemConfigs`**

- A mapping that associates chain IDs with their respective SystemConfig addresses.
- This will be used when updating dependencies along each chain.

**`dependencySet`**

- An `EnumerableSet` that stores the current list of chain IDs in the dependency set.
- This MUST replicate the same state as the one stored in the `L1BlockInterop` on L2 for each chain.

**`addChain`**

The `addChain` function adds a new chain to the op-governed cluster.

- The function MUST only be callable by the authorized `updater` role of the `SuperchainConfig`.
- The function MUST NOT add a chain ID to the dependency set if it is already included.
- The function MUST update dependencies to form a complete mesh graph,
  linking the new chain with all existing chains in the set.

> - The function MUST store the provided `SystemConfig` address in the systemConfigs mapping.

Reviewer 2: ```suggestion
- The function MUST store the provided `SystemConfig` address in the `systemConfigs` mapping.
```

> - The function MUST whitelist the new chain's portal in the `SharedLockbox`.

Reviewer 2: ```suggestion
- The function MUST allowlist the new chain's `OptimismPortal` in the `SharedLockbox`.
```

> - The function MUST emit the `ChainAdded` event.

Reviewer 2: with what arguments? I know it's seen in the example impl, but it's not formally defined if i didn't miss it, like other events

Author: good catch, i missed it

```solidity
function addChain(uint256 _chainId, address _systemConfig) external;
```

#### `OptimismPortal`

This contract will need an upgrade to integrate the `SharedLockbox` and start using the shared ETH liquidity.

**Interface and properties**

The `OptimismPortal` contract will add the following storage layout and modified functions:

**`SHARED_LOCKBOX`**

- An immutable address pointing to the `SharedLockbox` contract.

> - This address MUST immutable because all portals will point to the same lockbox, and this address SHOULD not change.

Reviewer 2: ```suggestion
- This address MUST be immutable because all `OptimismPortals` will point to the same `SharedLockbox`, and this address SHOULD not change.
```

**Integrating `SharedLockbox`**

The integration with the `SharedLockbox` involves adding extra steps when executing deposit transactions
or finalizing withdrawal transactions.

> These steps include locking and unlocking ETH without altering other aspects of the current portal implementation.

Reviewer 2: ```suggestion
These steps include locking and unlocking ETH without altering other aspects of the current `OptimismPortal` implementation.
```

Calls `unlockETH` on the `SharedLockbox` with the `tx.value`.

- The function MUST call `unlockETH` on the `SharedLockbox` if:
  - The token is `ETHER`.
  - `tx.value` is greater than zero.
- The function MUST NOT affect the handling of non-ETH tokens or existing verification logic.

## Reference implementation

### `SharedLockbox`

An example implementation could look like this:

```solidity
// OptimismPortals that are part of the dependency cluster.
mapping(address _portal => bool) internal _authorizedPortals;

function lockETH() external payable {
    require(_authorizedPortals[msg.sender], "Unauthorized");

    emit ETHLocked(msg.sender, msg.value);
}

function unlockETH(uint256 _value) external {
    require(_authorizedPortals[msg.sender], "Unauthorized");

    // Using `donateETH` to not trigger a deposit
    IOptimismPortal(msg.sender).donateETH{ value: _value }();

    emit ETHUnlocked(msg.sender, _value);
}

function authorizePortal(address _portal) external {
    require(msg.sender == superchainConfig, "Unauthorized");

    _authorizedPortals[_portal] = true;

    emit AuthorizedPortal(_portal);
}
```

### `LiquidityMigrator`

> A possible migrate function implementation can be:

Reviewer 3: NIT for consistency when talking about the implementations
```suggestion
A possible migrate function implementation could be:
```

    require(msg.sender == updater(), "Unauthorized");

    // Add to the dependency set and check it is not already added (`add()` returns false if it already exists)
    require(_dependencySet.add(_chainId), "Chain already added");

    // Store the system config
    systemConfigs[_chainId] = _systemConfig;

    // Loop through the dependency set and update the dependency for each chain
    for (uint256 i; i < _dependencySet.length() - 1; i++) {
        uint256 currentId = _dependencySet.at(i);

        // Skip recently added chain
        if (_chainId == currentId) continue;

        systemConfigs[currentId].addDependency(_chainId);

        systemConfigs[_chainId].addDependency(currentId);
    }

    address portal = _systemConfig.optimismPortal();

    // Authorize the portal on the shared lockbox
    SHARED_LOCKBOX.authorizePortal(portal);

    emit ChainAdded(_chainId, _systemConfig, portal);
}

function dependencySet() external view returns (uint256[] memory) {
    return dependencySet.values();
}

function isInDependencySet(uint256 _chainId) public view returns (bool) {
    return dependencySet.contains(_chainId);
}
```

### `OptimismPortal`

> Functions upgrades could look like this:

Reviewer 3: Leaving this jic you find better to avoid repeating 'look like this':
```suggestion
Functions upgrades could look as follow:
```

It provides two main functions: `lockETH` for depositing ETH into the lockbox,
and `unlockETH` for withdrawing ETH from the lockbox.
These functions are called by the `OptimismPortal` contracts to manage the shared ETH liquidity
when making deposits or finalizing withdrawals.

**Interface and properties:**

**`lockETH`**

Deposits and locks ETH into the lockbox's liquidity pool.

- The function MUST accept ETH.
- Only authorized `OptimismPortal` addresses MUST be allowed to interact.
- The function MUST emit the `ETHLocked` event with the correct `portal` and `amount`.

```solidity
function lockETH() external payable;
```

**`unlockETH`**

Withdraws a specified amount of ETH from the lockbox's liquidity pool.

- Only authorized `OptimismPortal` addresses MUST be allowed to interact.
- The function MUST emit the `ETHUnlocked` event with the correct `portal` and `amount`.

```solidity
function unlockETH(uint256 _value) external;
```

**`authorizePortal`**

Grants authorization to a specific `OptimismPortal` contract.

- Only `SuperchainConfig` address MUST be allowed to interact.
- The function MUST add the specified address to the mapping of authorized portals.

> - The function MUST emit the `AuthorizedPortal` event when a portal is successfully added.

Author: It is on the same section, can i link a non-header text?

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
**Table of Contents**

- [Overview](#overview)
- [Design](#design)
  - [Contracts introductions](#contracts-introductions)
    - [`SharedLockbox`](#sharedlockbox)
      - [**Interface and properties**](#interface-and-properties)
      - [**Events**](#events)
    - [`LiquidityMigrator`](#liquiditymigrator)
      - [**Interface and properties**](#interface-and-properties-1)
  - [Contracts upgrades](#contracts-upgrades)
    - [`SuperchainConfig`](#superchainconfig)
      - [**Interface and properties**](#interface-and-properties-2)
      - [**Events**](#events-1)
    - [`OptimismPortal`](#optimismportal)
      - [**Interface and properties**](#interface-and-properties-3)
- [Reference implementation](#reference-implementation)
  - [`SharedLockbox`](#sharedlockbox-1)
  - [`LiquidityMigrator`](#liquiditymigrator-1)
  - [`SuperchainConfig`](#superchainconfig-1)
  - [`OptimismPortal`](#optimismportal-1)

<!-- END doctoc generated TOC please keep comment here to allow auto update -->

## Overview

With interoperable ETH, withdrawals will fail if the referenced `OptimismPortal` lacks sufficient ETH.

> The `SharedLockbox` improves the Superchain's interoperable ETH withdrawal user experience and avoid this issue.

Reviewer 3: ```suggestion
The `SharedLockbox` improves the Superchain's interoperable ETH withdrawal user experience and avoids this issue.
```

  - The token is `ETHER`.
  - `msg.value` is greater than zero.

**`finalizeWithdrawalTransactionExternalProof`**

Calls `unlockETH` on the `SharedLockbox` with the `tx.value`.

- The function MUST call `unlockETH` on the `SharedLockbox` if:
  - The token is `ETHER`.
  - `tx.value` is greater than zero.

## Reference implementation

### `SharedLockbox`

An example implementation could look like this:

```solidity
// OptimismPortals that are part of the dependency cluster.
mapping(address _portal => bool) internal _authorizedPortals;

function lockETH() external payable {
    require(_authorizedPortals[msg.sender], "Unauthorized");

    emit ETHLocked(msg.sender, msg.value);
}

function unlockETH(uint256 _value) external {
    require(_authorizedPortals[msg.sender], "Unauthorized");

    // Using `donateETH` to not trigger a deposit
    IOptimismPortal(msg.sender).donateETH{ value: _value }();

    emit ETHUnlocked(msg.sender, _value);
}

function authorizePortal(address _portal) external {
    require(msg.sender == superchainConfig, "Unauthorized");

    _authorizedPortals[_portal] = true;

    emit AuthorizedPortal(_portal);
}
```

### `LiquidityMigrator`

> A possible migrate function implementation could be:

Reviewer 3: NIT: Not a strong opinion
```suggestion
A possible migration function implementation could be:
```

```solidity
// The Shared Lockbox contract
address internal immutable SHARED_LOCKBOX;

function migrateETH() external {
    SHARED_LOCKBOX.lockETH{ value: address(this).balance }();
}
```

### `SuperchainConfig`

> An example implementation could look as follow:

Reviewer 3: ```suggestion
An example implementation could look as follows:
```

```solidity
// The Shared Lockbox contract
address internal immutable SHARED_LOCKBOX;

// Mapping from chainId to SystemConfig address
mapping(uint256 _chainId => ISystemConfig) public systemConfigs;

> // Current dependency set list

Reviewer 3: This to avoid confusion since is not a list
```suggestion
// Current dependency set
```

and `unlockETH` for withdrawing ETH from the lockbox.
These functions are called by the `OptimismPortal` contracts to manage the shared ETH liquidity
when making deposits or finalizing withdrawals.

> These `OptimismPortals` will be allowlisted by the `SuperchanConfig` using the `authorizePortal` function

Reviewer 3: `OptimismPortal`s
